src/supervisor.tex

```diff
@@ -2261,3 +2281,108 @@ \chapter{``Svnapot'' Standard Extension for NAPOT Translation Contiguity, Versio
   types or accesses to invalid physical memory regions trigger page faults.
 \end{commentary}
+
+\chapter{``Svpbmt'' Standard Extension for Page-Based Memory Attributes, Version 0.1}
+\label{svpbmt}
+
+{\bf Warning! This draft specification is likely to change before being
+accepted as standard by the RISC-V Foundation.}
+
+In Sv39, Sv48, and Sv57, bits 62--61 of the page table entry indicate the use
+of page-based memory types that override the PMA(s) for the associated memory
+pages.  The encoding for the PBMT bits is captured in Table~\ref{pbmt}.
+
+\begin{table*}[h!]
+\begin{center}
+\begin{tabular}{|r|l|}
+\hline
+Value  & Page-Based Memory Attributes \\
+\hline
+0      & None \\
+1      & Non-cacheable, idempotent, weakly-ordered (RVWMO or RVTSO), main memory \\
+2      & Non-cacheable, non-idempotent, strongly-ordered (channel 0), I/O \\
+3      & {\em Reserved for future standard use} \\
+\hline
+\end{tabular}
+\end{center}
+\caption{Encodings for the PBMT field in Sv39, Sv48, and Sv57 PTEs.  Attributes
+not mentioned are inherited from the PMA associated with the physical address.}
+\label{pbmt}
+\end{table*}
+
+\begin{commentary}
+Future extensions may provide more and/or finer-grained control over which PMAs
+can be overridden.
+\end{commentary}
+
+With Svpbmt enabled, it is possible for multiple virtual aliases of the same
+physical page to exist simultaneously with different memory attributes.
+It is also possible for a U-mode or S-mode mapping through a PTE with Svpbmt
+enabled to observe different memory attributes for a given region of
+physical memory than a concurrent access to the same page performed by M-mode
+or when {\tt satp}.MODE=Bare.  In such cases, each individual access observes
+the memory attributes associated with its own path; aliases are not considered
+when determining attributes.
+
+\begin{commentary}
+For example, a cacheable access may be issued at the same time as a
+non-cacheable access to the same physical memory address.  In this case,
+if the former is performed first in the global memory order, then it will
+be evicted from the cache by the latter.  If on the other hand the cacheable
+access appears after the non-cacheable access, then the former may remain
+cached as it normally would.
+
+Likewise, accesses performed under memory indicating the non-idempotent
+attribute must not be merged with idempotent accesses to the same region
+in flight at the same time, as the non-idempotency of the former must
+be respected.  This is not expected to be a common situation.
+
+Note that Svpbmt cannot be used to completely prevent speculative reads from
+being performed to a region of memory for which the PMAs indicate idempotence,
+as speculation can still be performed via M-mode or via Bare mappings, which do
+not use the PBMTs.
+\end{commentary}
+
+Memory accesses are ordered according to the effective memory type for the
+address in question.  A region of memory which is considered main memory by the
```

Comment:
I can see how this paragraph (after the first sentence) may be a little confusing for some readers to parse correctly.  Maybe this and the following paragraph can all be condensed to:

The ordering of memory accesses by fence instructions is based on the effective memory type and its 'main memory' versus 'I/O' distinction.  This includes for the purposes of FENCE, .aq, and .rl.

A non-normative comment can provide any desired clarification(s).  Although it seems like the commentary in the two paragraphs all follows directly from the current arch specs and the above statement.


Memory accesses are ordered according to the effective memory type for the address in question.  If the address is

Reply:
P.S. These paragraphs (and some later text) tend to mix together two distinct forms of terminology:  Architectural "main memory" and "I/O" (as PMA attributes), and "main memory" and "I/O" as colloquial terms.  The former refer to how an access is handled or performed; the latter refer to the physical nature of what is at a physical address in a system.  For example, one can "map" an address with a "I/O" memory type to what physically is DRAM "main memory", or an address can have a "main memory" memory type but maps to a RAM buffer within an I/O device.

Reply:
There should be no imprecise colloquial usage, we should use proper definitions for these things.  I suggested "effective memory attributes" vs "physical memory attributes" above...if we go with that, this would become something like "If a memory access has an effective memory attribute of main memory, it obeys RVWMO (or RVTSO, if enabled).  If it has an effective memory attribute of I/O, it obeys I/O channel 0 ordering rules."